### DISCUSSION TITLE
numpy配列の引き算

### DISCUSSION THREAD

**Discussion by kioto** [kioto/BikeSimulation#3] [General]
Opus_Handle180Rotate.pyで以下のような記述がありますが、

```
     72 def calcRotCentVector(param1, param2):
     73     """回転中心ベクトル算出関数
     74     """
     75 
     76     # 回転中心ベクトルの算出
     77     n = [param1[0] - param2[0], param1[1] - param2[1], param1[2] - param2[2]]
     78     print("n = ", n)
     79     nn = param1 - param2
     80     print("nn = ", nn)
```

実はこれ、nとnnは同じになります。

```
>>> import numpy as np
>>> p1 = np.array([1/11,1/22])
>>> p1
array([0.09090909, 0.04545455])
>>> p2 = np.array([0.01, 0.02])
>>> p2
array([0.01, 0.02])
>>> p1[0] - p2[0]
0.08090909090909092
>>> p1 - p2
array([0.08090909, 0.02545455])
```

上記の動きを見ると、`p1 - p2`で後ろの方が切り捨てられてるように見えるんですが、これは表示の時だけの動作で、

```
>>> (p1 - p2)[0]
0.08090909090909092
```

と、要素を直接取り出すと同じ数値になります。

**Comment by ToshimitsuHIGASHI** [comment]
色々と恥ずかしいコードを見つけられて，嬉しいやら恥ずかしいやらです．
ご指摘頂いたnとnnは同じ処理をしているのは意識しておりました．
numpyの動きをイメージ出来ておらず，プログラムを作成している最中で
ベクトル計算が正しく理解しているかを試した痕跡になります．

Opusは完成版，Etudeは作成途中と位置づけており，本来ならばOpusに
した時点で消さなくてはいけないのですが，自己満足の塊でしかなかった
ので，そのままにしていました．

以上，言い訳になります．

ご指摘，有難うございました．

**Reply by kioto** [reply]
ここ、私も最初は「演算結果が違う？」と思ってちょっと悩んでました。
元のコードが、丸め誤差で悩んだ形跡かと思って色々と調べた記録を書いておきました。
